I wanted to make sure the build passed. After testing on a x86_64-linux machine I have a consistent failure with just one of the tests.
here is the stacktrace

```
gradio> test/test_examples.py xxxx.xxxxx.xx.x.....xx                             [ 47%]
gradio> test/test_external.py xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx..xxxxxx......... [ 58%]
gradio> ...xxx                                                                   [ 60%]
gradio> test/test_flagging.py xx.x.xx......                                      [ 63%]
gradio> test/test_interpretation.py .........                                    [ 65%]
gradio> test/test_mix.py .x..x                                                   [ 66%]
gradio> test/test_networking.py ..xx.x                                           [ 67%]
gradio> test/test_pipelines.py x                                                 [ 68%]
gradio> test/test_processing_utils.py .x.x..............x..........x             [ 75%]
gradio> test/test_queueing.py ...............xxx.x....                           [ 80%]
gradio> test/test_reload.py .                                                    [ 81%]
gradio> test/test_routes.py xxxxxxxxxxxxxx.xx...xxxxxxxx                         [ 87%]
gradio> test/test_tunneling.py x                                                 [ 87%]
gradio> test/test_utils.py .xxx....x.......x........xxxxxxxxxxx...xx...........  [100%]
gradio>
gradio> =================================== FAILURES ===================================
gradio> __________________________ TestJSON.test_in_interface __________________________
gradio>
gradio> x = array(['MM'], dtype=object)
gradio>
gradio>     def _ensure_numeric(x):
gradio>         if isinstance(x, np.ndarray):
gradio>             if is_integer_dtype(x) or is_bool_dtype(x):
gradio>                 x = x.astype(np.float64)
gradio>             elif is_object_dtype(x):
gradio>                 try:
gradio> >                   x = x.astype(np.complex128)
gradio> E                   ValueError: complex() arg is a malformed string
gradio>
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>
gradio> During handling of the above exception, another exception occurred:
gradio>
gradio> x = array(['MM'], dtype=object)
gradio>
gradio>     def _ensure_numeric(x):
gradio>         if isinstance(x, np.ndarray):
gradio>             if is_integer_dtype(x) or is_bool_dtype(x):
gradio>                 x = x.astype(np.float64)
gradio>             elif is_object_dtype(x):
gradio>                 try:
gradio>                     x = x.astype(np.complex128)
gradio>                 except (TypeError, ValueError):
gradio>                     try:
gradio> >                       x = x.astype(np.float64)
gradio> E                       ValueError: could not convert string to float: 'MM'
gradio>
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>
gradio> The above exception was the direct cause of the following exception:
gradio>
gradio> self = <test.test_components.TestJSON object at 0x7fff767a4670>
gradio>
gradio>     @pytest.mark.asyncio
gradio>     async def test_in_interface(self):
gradio>         """
gradio>         Interface, process
gradio>         """
gradio>
gradio>         def get_avg_age_per_gender(data):
gradio>             print(data)
gradio>             return {
gradio>                 "M": int(data[data["gender"] == "M"].mean()),
gradio>                 "F": int(data[data["gender"] == "F"].mean()),
gradio>                 "O": int(data[data["gender"] == "O"].mean()),
gradio>             }
gradio>
gradio>         iface = gr.Interface(
gradio>             get_avg_age_per_gender,
gradio>             gr.Dataframe(headers=["gender", "age"]),
gradio>             "json",
gradio>         )
gradio>         y_data = [
gradio>             ["M", 30],
gradio>             ["F", 20],
gradio>             ["M", 40],
gradio>             ["O", 20],
gradio>             ["F", 30],
gradio>         ]
gradio> >       assert (
gradio>             await iface.process_api(
gradio>                 0, [{"data": y_data, "headers": ["gender", "age"]}], state={}
gradio>             )
gradio>         )["data"][0] == {
gradio>             "M": 35,
gradio>             "F": 25,
gradio>             "O": 20,
gradio>         }
gradio>
gradio> test/test_components.py:1746:
gradio> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
gradio> gradio/blocks.py:1032: in process_api
gradio>     result = await self.call_function(
gradio> gradio/blocks.py:844: in call_function
gradio>     prediction = await anyio.to_thread.run_sync(
gradio> /nix/store/m8ix667n94csdc81m5aq55kfypc1bn3c-python3.10-anyio-3.7.0/lib/python3.1
gradio>     return await get_asynclib().run_sync_in_worker_thread(
gradio> /nix/store/m8ix667n94csdc81m5aq55kfypc1bn3c-python3.10-anyio-3.7.0/lib/python3.1
gradio>     return await future
gradio> /nix/store/m8ix667n94csdc81m5aq55kfypc1bn3c-python3.10-anyio-3.7.0/lib/python3.1
gradio>     result = context.run(func, *args)
gradio> test/test_components.py:1729: in get_avg_age_per_gender
gradio>     "M": int(data[data["gender"] == "M"].mean()),
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     return NDFrame.mean(self, axis, skipna, numeric_only, **kwargs)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     return self._stat_function(
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     return self._reduce(
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     res = df._mgr.reduce(blk_func)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     nbs = blk.reduce(func)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     result = func(self.values)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     return op(values, axis=axis, skipna=skipna, **kwds)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     return f(*args, **kwargs)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     result = alt(values, axis=axis, skipna=skipna, **kwds)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     result = func(values, axis=axis, skipna=skipna, mask=mask, **kwargs)
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.
gradio>     the_sum = _ensure_numeric(values.sum(axis, dtype=dtype_sum))
gradio> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
gradio>
gradio> x = array(['MM'], dtype=object)
gradio>
gradio>     def _ensure_numeric(x):
gradio>         if isinstance(x, np.ndarray):
gradio>             if is_integer_dtype(x) or is_bool_dtype(x):
gradio>                 x = x.astype(np.float64)
gradio>             elif is_object_dtype(x):
gradio>                 try:
gradio>                     x = x.astype(np.complex128)
gradio>                 except (TypeError, ValueError):
gradio>                     try:
gradio>                         x = x.astype(np.float64)
gradio>                     except ValueError as err:
gradio>                         # GH#29941 we get here with object arrays containing strs
gradio> >                       raise TypeError(f"Could not convert {x} to numeric") from err
gradio> E                       TypeError: Could not convert ['MM'] to numeric
gradio>
gradio> /nix/store/0a4c7ypszwhm3zz3hs5qizjkjyq0jqc3-python3.10-pandas-2.0.3/lib/python3.10/site-packages/pandas/core/nanops.py:1686: TypeError
gradio> ----------------------------- Captured stdout call -----------------------------
gradio>   gender  age
gradio> 0      M   30
gradio> 1      F   20
gradio> 2      M   40
gradio> 3      O   20
gradio> 4      F   30
```

Nothing jumps at me in particular, so I would advise just disabling the test. Let me know if you disagree.

(in the process of testing I discovered several failures with upstream dependencies that I have reported. Note where related to the changes made here.)

--- a/pkgs/development/python-modules/ffmpy/default.nix
+++ b/pkgs/development/python-modules/ffmpy/default.nix
@@ -1,0 +1,55 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, pythonOlder
+, pytestCheckHook
+, go
+, ffmpeg-headless
+}:
+
+buildPythonPackage rec {
+  pname = "ffmpy";
+  version = "0.3.1";
+  format = "setuptools";
+
+  disabled = pythonOlder "3.6";
+
+  src = fetchFromGitHub {
+    owner = "Ch00k";
+    repo = "ffmpy";
+    rev = "refs/tags/${version}";
+    hash = "sha256-kuLhmCG80BmXdqpW67UanBnuYiL2Oh1jKt7IgmVNEAM=";
+  };
+
+  postPatch = ''
+    # default to store ffmpeg
+    substituteInPlace ffmpy.py \
+      --replace 'executable="ffmpeg",' 'executable="${ffmpeg-headless}/bin/ffmpeg",'
+
+    #  The tests test a mock that does not behave like ffmpeg. If we default to the nix-store ffmpeg they fail.
+    for fname in tests/*.py; do
+      echo 'FFmpeg.__init__.__defaults__ = ("ffmpeg", *FFmpeg.__init__.__defaults__[1:])' >>"$fname"
+    done
+  '';
+
+  pythonImportsCheck = [ "ffmpy" ];
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    go
+  ];
+
+  # the vendored ffmpeg mock binary assumes FHS
+  preCheck = ''
+    rm -v tests/ffmpeg/ffmpeg
+    HOME=$(mktemp -d) go build -o ffmpeg tests/ffmpeg/ffmpeg.go
+    export PATH=".:$PATH"
+  '';
+
+  meta = with lib; {
+    description = "A simple python interface for FFmpeg/FFprobe";
+    homepage = "https://github.com/Ch00k/ffmpy";
+    license = licenses.mit;
+    maintainers = with maintainers; [ pbsds ];
+  };
+}

--- a/pkgs/development/python-modules/gradio/conftest-skip-network-errors.py
+++ b/pkgs/development/python-modules/gradio/conftest-skip-network-errors.py
@@ -1,0 +1,57 @@
+# This is a pytest hook that skips tests that tries to access the network.
+# These tests will immediately fail, then get marked as "Expected fail" (xfail).
+
+from _pytest.runner import pytest_runtest_makereport as orig_pytest_runtest_makereport
+
+# We use BaseException to minimize the chance it gets caught and 'pass'ed
+class NixNetworkAccessDeniedError(BaseException):
+    pass
+
+def pytest_runtest_makereport(item, call):
+    """
+    Modifies test results after-the-fact. The function name is magic, see:
+    https://docs.pytest.org/en/7.1.x/reference/reference.html?highlight=pytest_runtest_makereport#std-hook-pytest_runtest_makereport
+    """
+
+    def iterate_exc_chain(exc: Exception):
+        """
+        Recurses through exception chain, yielding all exceptions
+        """
+        yield exc
+        if getattr(exc, "__context__", None) is not None:
+            yield from iterate_exc_chain(exc.__context__)
+        if getattr(exc, "__cause__", None) is not None:
+            yield from iterate_exc_chain(exc.__cause__)
+
+    tr = orig_pytest_runtest_makereport(item, call)
+    if call.excinfo is not None:
+        for exc in iterate_exc_chain(call.excinfo.value):
+            if isinstance(exc, NixNetworkAccessDeniedError):
+                tr.outcome, tr.wasxfail = 'skipped', "reason: Requires network access."
+            if isinstance(exc, FileNotFoundError):  # gradio specific
+                tr.outcome, tr.wasxfail = 'skipped', "reason: Pypi dist bad."
+    return tr
+
+# replace network access with exception
+
+def deny_network_access(*a, **kw):
+    raise NixNetworkAccessDeniedError
+
+import httpx
+import requests
+import socket
+import urllib
+import urllib3
+import websockets
+
+httpx.AsyncClient.get = deny_network_access
+httpx.AsyncClient.head = deny_network_access
+httpx.Request = deny_network_access
+requests.get = deny_network_access
+requests.head = deny_network_access
+requests.post = deny_network_access
+socket.socket.connect = deny_network_access
+urllib.request.Request = deny_network_access
+urllib.request.urlopen = deny_network_access
+urllib3.connection.HTTPSConnection._new_conn = deny_network_access
+websockets.connect = deny_network_access

--- a/pkgs/development/python-modules/gradio/default.nix
+++ b/pkgs/development/python-modules/gradio/default.nix
@@ -1,0 +1,175 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+, pythonOlder
+, pythonRelaxDepsHook
+, writeText
+
+# pyproject
+, hatchling
+, hatch-requirements-txt
+, hatch-fancy-pypi-readme
+
+# runtime
+, setuptools
+, aiofiles
+, aiohttp
+, altair
+, fastapi
+, ffmpy
+, markdown-it-py
+, mdit-py-plugins
+, markupsafe
+, matplotlib
+, numpy
+, orjson
+, pandas
+, pillow
+, pycryptodome
+, python-multipart
+, pydub
+, pyyaml
+, requests
+, uvicorn
+, jinja2
+, fsspec
+, httpx
+, pydantic
+, websockets
+, typing-extensions
+
+# check
+, pytestCheckHook
+, pytest-asyncio
+, mlflow
+, huggingface-hub
+, transformers
+, wandb
+, respx
+, scikit-image
+, ipython
+, ffmpeg
+, vega_datasets
+, boto3
+}:
+
+buildPythonPackage rec {
+  pname = "gradio";
+  version = "3.20.1";
+  format = "pyproject";
+
+  disabled = pythonOlder "3.7";
+
+  # We use the Pypi release, as it provides prebuilt webui assets,
+  # and has more frequent releases compared to github tags
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-oG97GwehyBWjWXzDqyfj+x2mAfM6OQhYKdA3j0Rv8Vs=";
+  };
+
+  pythonRelaxDeps = [
+    "mdit-py-plugins"
+  ];
+
+  nativeBuildInputs = [
+    pythonRelaxDepsHook
+    hatchling
+    hatch-requirements-txt
+    hatch-fancy-pypi-readme
+  ];
+
+  propagatedBuildInputs = [
+    setuptools # needs pkg_resources
+    aiofiles
+    aiohttp
+    altair
+    fastapi
+    ffmpy
+    markdown-it-py
+    mdit-py-plugins
+    markupsafe
+    matplotlib
+    numpy
+    orjson
+    pandas
+    pillow
+    pycryptodome
+    python-multipart
+    pydub
+    pyyaml
+    requests
+    uvicorn
+    jinja2
+    fsspec
+    httpx
+    pydantic
+    websockets
+    typing-extensions
+  ] ++ markdown-it-py.optional-dependencies.linkify;
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    pytest-asyncio
+    mlflow
+    #comet-ml # FIXME: enable once packaged
+    huggingface-hub
+    transformers
+    wandb
+    respx
+    scikit-image
+    ipython
+    ffmpeg
+    vega_datasets
+    boto3
+    # shap is needed as well, but breaks too often
+  ];
+
+  # Add a pytest hook skipping tests that access network, marking them as "Expected fail" (xfail).
+  # We additionally xfail FileNotFoundError, since the gradio devs often fail to upload test assets to pypi.
+  preCheck = let
+  in ''
+    export HOME=$TMPDIR
+    cat ${./conftest-skip-network-errors.py} >> test/conftest.py
+  '';
+
+  disabledTests = [
+    # Actually broken
+    "test_mount_gradio_app"
+
+    # FIXME: enable once comet-ml is packaged
+    "test_inline_display"
+    "test_integration_comet"
+
+    # Flaky, tries to pin dependency behaviour. Sensitive to dep versions
+    # These error only affect downstream use of the check dependencies.
+    "test_no_color"
+    "test_in_interface_as_output"
+    "test_should_warn_url_not_having_version"
+
+    # Flaky, unknown reason
+    "test_in_interface"
+
+    # shap is too often broken in nixpkgs
+    "test_shapley_text"
+  ];
+  disabledTestPaths = [
+    # makes pytest freeze 50% of the time
+    "test/test_interfaces.py"
+  ];
+  #pytestFlagsArray = [ "-x" "-W" "ignore" ]; # uncomment for debugging help
+
+  # check the binary works outside the build env
+  doInstallCheck = true;
+  postInstallCheck = ''
+    env --ignore-environment $out/bin/gradio --help >/dev/null
+  '';
+
+  pythonImportsCheck = [ "gradio" ];
+
+  meta = with lib; {
+    homepage = "https://www.gradio.app/";
+    description = "Python library for easily interacting with trained machine learning models";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ pbsds ];
+  };
+}

--- a/pkgs/development/python-modules/markdown-it-py/default.nix
+++ b/pkgs/development/python-modules/markdown-it-py/default.nix
@@ -19,6 +19,7 @@
 , stdenv
 , pytest-regressions
 , pytestCheckHook
+, pythonRelaxDepsHook
 , pythonOlder
 }:
 
@@ -36,7 +37,13 @@ buildPythonPackage rec {
     hash = "sha256-qdRU1BxczFDGoIEtl0ZMkKNn4p5tec8YuPt5ZwX5fYM=";
   };
 
+  # fix downstrem usage of markdown-it-py[linkify]
+  pythonRelaxDeps = [
+    "linkify-it-py"
+  ];
+
   nativeBuildInputs = [
+    pythonRelaxDepsHook
     flit-core
   ];
 

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -3674,6 +3674,8 @@ self: super: with self; {
 
   ffmpeg-progress-yield = callPackage ../development/python-modules/ffmpeg-progress-yield { };
 
+  ffmpy = callPackage ../development/python-modules/ffmpy { };
+
   fiblary3-fork = callPackage ../development/python-modules/fiblary3-fork { };
 
   fido2 = callPackage ../development/python-modules/fido2 { };

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -4476,6 +4476,8 @@ self: super: with self; {
 
   gradient_statsd = callPackage ../development/python-modules/gradient_statsd { };
 
+  gradio = callPackage ../development/python-modules/gradio { };
+
   grammalecte = callPackage ../development/python-modules/grammalecte { };
 
   grandalf = callPackage ../development/python-modules/grandalf { };

